[AG-9792] Feature(cell-editing): allow async loading of matches when typing in the agrichselectcelleditor input

sdwvit opened the pull request:
feat(rich-select): add support for async filtering with server-driven results

 - Introduce `filterListAsync` flag to enable server-side filtering via async values callback
 - Add `onSearch` callback with debounce (300ms default) for real-time search requests
 - Update `RichCellEditorValuesCallbackParams` to include search string parameter
 - Enhance `setValueList` to handle Promise-based value lists and loading states
 - Display loading indicator and handle race conditions in async search requests
 - Add error warning when `filterListAsync` is used without `filterList=true`
 - Improve UX with dynamic state handling: loading, no matches, and ready states
 - Update documentation with async filtering example and API reference
 - Increase module size for `RichSelectModule` due to new async filtering logic

Fix [AG-9792](https://ag-grid.atlassian.net/browse/AG-9792)

<img width="310" height="138" alt="Screenshot 2025-11-11 at 10 45 32" src="https://github.com/user-attachments/assets/45896ea6-9115-44de-bf44-4f134c64327a" />
<img width="421" height="232" alt="Screenshot 2025-11-11 at 10 45 23" src="https://github.com/user-attachments/assets/acd38931-1489-4d36-a3f4-8f27c5d84051" />

seanlandsman committed:
DASH-11 Auto grid bump and publish trigger

seanlandsman committed:
DASH-11 Auto grid bump and publish trigger

seanlandsman committed:
DASH-11 Auto grid bump and publish trigger

sdwvit committed:
Merge branch 'latest' into sdwvit/feature/cell-editing/allow-async-loading-of-matches-when-typing-in-the-agrichselectcelleditor-input/AG-9792

# Conflicts:
#	.github/workflows/beta-publish.yml

sdwvit committed:
Allow asynchronous loading of matches in agRichSelectCellEditor

 - Added onSearch callback to RichSelectParams for async search handling.
 - Updated agRichSelect to call onSearch when filtering values.
 - Introduced filterListAsync option to enable async filtering behavior.
 - Enhanced error messages for filterListAsync usage in errorText.
 - Debounced onSearch callback to improve performance during typing.

sdwvit committed:
Implement async loading for value lists in agRichSelectCellEditor

 - Add setValueListAsync method to handle async value list loading.
 - Show loading overlay while fetching values and hide it afterward.
 - Update onValuesPromise to manage async value list setting.
 - Refactor onSearchCallback to support async value fetching.

sdwvit committed:
no need for the check if we return early

sdwvit committed:
expose debounce delay

StephenCooper commented on packages/ag-grid-community/src/interfaces/iRichCellEditorParams.ts:
```diff
@@ -7,24 +7,27 @@ export interface IRichCellEditorRendererParams<TValue> extends ICellEditorRender
 
-export interface RichSelectParams<TValue = any> extends AgPickerFieldParams<AgComponentSelectorType> {
+export interface RichSelectParams<TValue = any>
+    extends AgPickerFieldParams<AgComponentSelectorType>,
```
Do we need this complication of the type definition?

StephenCooper commented on packages/ag-grid-community/src/interfaces/iRichCellEditorParams.ts:
```diff
@@ -56,6 +59,13 @@ export interface IRichCellEditorParams<TData = any, TValue = any, GValue = any>
+     * When `filterList` is `true`, setting this to `true` and user typing will
+     * call the `values` function to fetch filtered results asynchronously.
+     * @default false
+     */
```
Can we update the copy.

StephenCooper commented on packages/ag-grid-community/src/validation/errorMessages/errorText.ts:
```diff
@@ -719,6 +719,8 @@ export const AG_GRID_ERRORS = {
     292: ({ colId }: { colId: string }) =>
         `The Multi Filter for column '${colId}' has buttons configured against the child filters. When 'enableFilterHandlers=true', buttons must instead be provided against the parent Multi Filter params. The child filter buttons will be ignored.` as const,
+    293: () =>
+        `Setting \`filterListAsync\` to \`true\` in agRichSelectCellEditor has no effect when \`values()\` returns non-Promise result or \`filterList\` set to \`false\`.` as const,
```
Please update

gportela85 reviewed (approved):
In general these changes are fine, we just need to validate that these properties
```
Pick<
            IRichCellEditorParams<any, TValue>,
            | 'filterList'
            | 'searchType'
            | 'highlightMatch'
            | 'multiSelect'
            | 'suppressDeselectAll'
            | 'suppressMultiSelectPillRenderer'
            | 'allowTyping'
        > {
```

will show up in the docs.

sdwvit committed:
Refine error messages and enhance RichSelectParams interface

 - Update error message for `filterListAsync` to clarify its relevance.
 - Simplify `RichSelectParams` interface by removing redundant properties.
 - Add detailed comments for `filterListAsync` to explain its usage.

sdwvit committed:
bundle size

StephenCooper committed:
Update jsdoc

sdwvit committed:
Merge branch 'latest' into sdwvit/feature/cell-editing/allow-async-loading-of-matches-when-typing-in-the-agrichselectcelleditor-input/AG-9792

sdwvit committed:
Merge remote-tracking branch 'origin/sdwvit/feature/cell-editing/allow-async-loading-of-matches-when-typing-in-the-agrichselectcelleditor-input/AG-9792' into sdwvit/feature/cell-editing/allow-async-loading-of-matches-when-typing-in-the-agrichselectcelleditor-input/AG-9792

sdwvit committed:
Add support for showing no results message in agRichSelect when allowNoResultsCopy is enabled

sdwvit committed:
- Update documentation to reflect new async and filtering capabilities

sdwvit committed:
pr comments

sdwvit committed:
pr comments

sdwvit committed:
Merge remote-tracking branch 'origin/sdwvit/feature/cell-editing/allow-async-loading-of-matches-when-typing-in-the-agrichselectcelleditor-input/AG-9792' into sdwvit/feature/cell-editing/allow-async-loading-of-matches-when-typing-in-the-agrichselectcelleditor-input/AG-9792

sdwvit committed:
Update rich select cell editor to handle async values and improve loading state management

 - Added isAsync flag to RichSelectParams to track async value loading state
 - Improved async value handling: setValueList now supports Promise<TValue[]> directly
 - Modified setValueList to handle undefined valueList for clearing previous results
 - Enhanced loading state management: show loading while async values are pending
 - Fixed no-results handling: show 'No matches' label only when appropriate
 - Removed redundant onValuesPromise logic; replaced with direct setValueList calls
 - Added proper null/undefined checks for valueList to prevent incorrect state
 - Updated onSearchCallback to handle async values correctly and avoid invalid calls
 - Improved type safety by ensuring params passed to values callback include search field

sdwvit committed:
wip stuff

Commit:
update the doc

Signed-off-by: Victor Musienko <victor.m@noibu.com>

Commit:
update the doc

Signed-off-by: Victor Musienko <victor.m@noibu.com>

Commit:
fix readme

Signed-off-by: Victor Musienko <victor.m@noibu.com>

sdwvit committed:
Simplify loading state management and improve async value handling in rich select components

sdwvit committed:
Improve async search handling and fix race conditions in rich select components

 - Added currentSearchRequest counter to track and discard outdated search results
 - Updated onSearchCallback to use request counter for avoiding stale async updates
 - Modified setValueList to properly handle Promise<TValue[]> with result validation
 - Renamed toggleVisibility to accept forceVisible flag for clearer intent
 - Updated hideLoadingStateComp to hideStateComp for better method naming
 - Improved list visibility logic to properly handle empty lists and states
 - Enhanced loading state management to prevent invalid state transitions

sdwvit committed:
Merge branch 'latest' into sdwvit/feature/cell-editing/allow-async-loading-of-matches-when-typing-in-the-agrichselectcelleditor-input/AG-9792

# Conflicts:
#	packages/ag-grid-community/src/validation/errorMessages/errorText.ts
#	packages/ag-grid-enterprise/src/widgets/agRichSelectList.ts
#	testing/module-size/moduleDefinitions.ts
#	testing/shared/moduleDefinitions.ts

sdwvit committed:
merge

sdwvit committed:
Refactor rich select placeholder logic and state management

 - Extracted placeholder text logic into getPlaceholderText() method with i18n fallbacks
 - Added isFullAsync() helper to determine when full async mode is active
 - Replaced direct placeholder assignment with dynamic placeholder from getPlaceholderText()
 - Introduced explicit state type aliases and constants for clearer state handling
 - Moved state management to public methods with improved visibility control
 - Simplified state transition logic using named state constants
 - Enhanced type safety with explicit state type definitions
 - Removed redundant private state management methods

sdwvit committed:
Fix incorrect list update logic and improve async handling in rich select

 - Removed redundant setCurrentList call when list is already current
 - Added null/undefined check for getCurrentList to prevent potential errors
 - Updated values assignment to occur only when valueList has changed
 - Simplified refresh logic by removing redundant conditionals
 - Improved async state handling with proper valueList comparison

Commit:
Allow async loading of matches in agrichselectcelleditor input

 - Enable searching without input field for historical reasons
 - Show picker when valueList is a promise
 - Maintain existing behavior for non-async editing

Signed-off-by: Victor Musienko <victor.m@noibu.com>

Commit:
- Refactor loading state management for async value lists
 - Simplify state handling logic in agRichSelect and agRichSelectList
 - Ensure picker visibility is consistent with search input
 - Improve handling of current search requests to avoid stale results

Signed-off-by: Victor Musienko <victor.m@noibu.com>

Commit:
Merge branch 'refs/heads/latest' into sdwvit/feature/cell-editing/allow-async-loading-of-matches-when-typing-in-the-agrichselectcelleditor-input/AG-9792

Commit:
support loading spinner and refactor label switching

Signed-off-by: Victor Musienko <victor.m@noibu.com>

Commit:
move debounce logic inside agRichSelect.ts

Signed-off-by: Victor Musienko <victor.m@noibu.com>

Commit:
rename states

Signed-off-by: Victor Musienko <victor.m@noibu.com>

Commit:
fix semi-async mode losing values

Signed-off-by: Victor Musienko <victor.m@noibu.com>

Commit:
idea files

Signed-off-by: Victor Musienko <victor.m@noibu.com>

StephenCooper commented on packages/ag-grid-community/src/interfaces/iRichCellEditorParams.ts:
```diff
@@ -30,10 +31,17 @@ export interface RichSelectParams<TValue = any> extends AgPickerFieldParams<AgCo
 
 export interface RichCellEditorValuesCallback<TData = any, TValue = any> {
-    (params: ICellEditorParams<TData, TValue>): TValue[] | Promise<TValue[]>;
+    (params: RichCellEditorValuesCallbackParams<TData, TValue>): TValue[] | Promise<TValue[]>;
```
Is this correct? Does RichSelectParams have what ICellEditorParams does?

sdwvit replied:
yes, code expects RichCellEditorParams which this new type extends from.

StephenCooper commented on packages/ag-grid-enterprise/src/richSelect/richSelectCellEditor.ts:
```diff
@@ -14,35 +15,39 @@ export class RichSelectCellEditor<TData = any, TValue = any, TContext = any> ext
                 }
 
-                this.processEventKey(eventKey);
+                this.processEventKey(this.params.eventKey);
```
Can this be pulled off `this.params` like it was before?

Commit:
- Remove isAsync property from RichSelectParams interface

Signed-off-by: Victor Musienko <victor.m@noibu.com>

StephenCooper commented on packages/ag-grid-community/src/validation/errorMessages/errorText.ts:
```diff
@@ -721,6 +721,8 @@ export const AG_GRID_ERRORS = {
     293: () =>
         `The grid was initialised detached from the DOM and was then inserted into a Shadow Root. Theme styles are probably broken. Pass the themeStyleContainer grid option to let the grid know where in the document to insert theme CSS.` as const,
+    294: () =>
+        `When using the \`agRichSelectCellEditor\` setting \`filterListAsync = true\` requires \`filterList = true\` and the \`values()\` callback must return a Promise of filtered values.` as const,
```
This needs updating to include allowTyping

StephenCooper commented on packages/ag-grid-enterprise/src/richSelect/richSelectCellEditor.ts:
```diff
@@ -75,17 +100,17 @@ export class RichSelectCellEditor<TData = any, TValue = any, TContext = any> ext
             suppressDeselectAll,
             suppressMultiSelectPillRenderer,
-        } = params;
+        } = this.params;
```
prefer `params` over `this.params`

StephenCooper commented on packages/ag-grid-enterprise/src/richSelect/richSelectCellEditor.ts:
```diff
@@ -101,37 +126,82 @@ export class RichSelectCellEditor<TData = any, TValue = any, TContext = any> ext
-        let valuesPromise;
+        let valueList;
+
+        if (filterListAsync && !filterList) {
```
Do we want to be consistent with `allowTyping` as a requirement?

sdwvit replied:
good point, updated

StephenCooper commented on packages/ag-grid-enterprise/src/richSelect/richSelectCellEditor.ts:
```diff
@@ -101,37 +126,82 @@ export class RichSelectCellEditor<TData = any, TValue = any, TContext = any> ext
+        }
+
+        if (typeof params.values !== 'function') {
+            // potentially allow sync values too
```
Is this dead code based on the documented restrictions for this feature?

sdwvit replied:
it's not dead, it's an early exit, but yes. In case we want to make 'values' dynamic, we can by easily extending the code here.

StephenCooper commented on documentation/ag-grid-docs/src/content/docs/provided-cell-editors-rich-select/index.mdoc:
```diff
@@ -187,14 +190,16 @@ columnDefs: [
 ## Async Values
 
-List values can be provided asynchronously to the editor as shown below:
+The `agRichSelectCellEditor` supports asynchronous loading of dropdown values, enabling efficient handling of large datasets and server-driven filtering. This feature is ideal for applications requiring dynamic, real-time filtering without loading all data upfront.
```
can you swap all "server-driven" with "server side" I think this is more in line with other places in our docs.

StephenCooper commented on packages/ag-grid-enterprise/src/widgets/agRichSelectList.ts:
```diff
@@ -80,6 +98,42 @@ export class AgRichSelectList<TValue, TEventType extends string = AgRichSelectLi
+        }
+        if (loadingState === STATE_LOADING) {
+            eStateCompLabel.textContent = loadingLabel;
+            eLoadingIcon?.classList?.toggle('ag-hidden', false);
```
We have a `_setDisplayed` utility which might be more appropriate to use

sdwvit committed:
Update rich select async filtering requirements and improve documentation

 - Added allowTyping = true requirement in error message for filterListAsync
 - Updated documentation to reflect server-side filtering terminology
 - Modified rich select list visibility logic to use _setDisplayed helper
 - Added new Nx dev fast run configuration for quicker development builds
 - Updated example to match new server-side filtering terminology

StephenCooper commented on documentation/ag-grid-docs/src/content/docs/provided-cell-editors-rich-select/_examples/rich-select-async-values/main.ts:
```diff
@@ -22,9 +22,13 @@ function getRandomNumber(min: number, max: number) {
     return new Promise((resolve) => {
-        setTimeout(() => resolve(languages), 1000);
+        setTimeout(() => {
+            console.log('Grid requested `' + search + '` from server.');
```
Does this console log make sense for the normal async case?

sdwvit replied:
It doesn't, I think it's fine since the second column is full async. Useful to see a difference.

StephenCooper commented on documentation/ag-grid-docs/src/content/docs/provided-cell-editors-rich-select/index.mdoc:
```diff
@@ -204,15 +209,53 @@ function getValueFromServer(_params: ICellEditorParams): Promise<string[]> {
 
+### Server Side Filtering
+
+For advanced filtering scenarios, combine async `values` callback, `allowTyping`, `filterList`, and `filterListAsync` to enable real-time, server-side filtering.
```
```suggestion
For advanced filtering scenarios, combine async `values` callback, `allowTyping`, `filterList`, and `filterListAsync` to enable server-side filtering.
```

StephenCooper committed:
Update documentation/ag-grid-docs/src/content/docs/provided-cell-editors-rich-select/index.mdoc
